### PR TITLE
Edit phone screen (Part 4) - Refactoring and test cases 

### PR DESCRIPTION
**Pull request by BradleyWire**
### Issues

In this part i've refactored to downsize the UserDataSource, update the test cases for phone number  changes and also added the confirmation behaviour to navigate to the new "verify phone" screen. 
#### APK
[Download build #1177](http://10.10.124.11:8080/job/Pull%20Request%20Builder/1177/artifact/build/artifact/wire-dev-PR2587-1177.apk)

### CODE REVIEW

File: app/src/main/AndroidManifest.xml [hunk 1]
```diff
@@ -227,7 +227,14 @@
             android:name=".conversation.folders.moveto.MoveToFolderActivity"
             android:screenOrientation="portrait" />
         <activity android:name=".settings.devices.detail.SettingsDeviceDetailActivity" />
-      
+
+        <activity
+            android:name="com.waz.zclient.settings.account.editphonenumber.EditPhoneNumberActivity"
+            android:hardwareAccelerated="true"
```
Comment (mejdoo) [inline app/src/main/AndroidManifest.xml]: I'm just wondering why we need to specify explicitly `android:hardwareAccelerated="true"` here? You you have some heathy tasks ?

File: app/src/main/kotlin/com/waz/zclient/settings/account/editphonenumber/SettingsAccountPhoneNumberViewModel.kt [hunk 2]
```diff
@@ -17,31 +18,30 @@ import com.waz.zclient.user.domain.usecase.phonenumber.PhoneNumberInvalid
...
-
 class SettingsAccountPhoneNumberViewModel(
     private val validatePhoneNumberUseCase: ValidatePhoneNumberUseCase,
     private val changePhoneNumberNumberUseCase: ChangePhoneNumberUseCase,
     private val countryCodeAndPhoneNumberUseCase: CountryCodeAndPhoneNumberUseCase,
     private val deletePhoneNumberUseCase: DeletePhoneNumberUseCase
 ) : ViewModel() {
 
-    private var _countryCodeErrorLiveData = MutableLiveData<CountryCodeErrorMessage>()
+    private var _countryCodeErrorLiveData = MutableLiveData<PhoneNumberErrorMessage>()
```
Comment (mejdoo) [inline app/src/main/kotlin/com/waz/zclient/settings/account/editphonenumber/SettingsAccountPhoneNumberViewModel.kt]: Is there any way to reduce the number of LiveData objects? It seems like we have a huge list in this viewModel.

File: app/src/main/kotlin/com/waz/zclient/user/data/phone/PhoneNumberDataSource.kt [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+import com.waz.zclient.user.data.source.local.UsersLocalDataSource
+import com.waz.zclient.user.data.source.remote.UsersRemoteDataSource
+import kotlinx.coroutines.runBlocking
+
+class PhoneNumberDataSource(
+    private val usersRemoteDataSource: UsersRemoteDataSource,
+    private val usersLocalDataSource: UsersLocalDataSource
+) : PhoneNumberRepository {
+
+    override suspend fun changePhone(phone: String) = changePhoneRemotely(phone)
```
Comment (mejdoo) [inline app/src/main/kotlin/com/waz/zclient/user/data/phone/PhoneNumberDataSource.kt]: I guess you shouldn't use `usersRemoteDataSource` `usersLocalDataSource` here. Either you create `PhoneNumberRemoteDataSource` and `PhoneNumberLocalDataSource` or you keep all those methods in `UsersDataSource`

Comment (BradleyWire) [reply app/src/main/kotlin/com/waz/zclient/user/data/phone/PhoneNumberDataSource.kt]: The way I see it is: the `UserLocalDataSource` hit the Users table in the database and `UserRemoteDataSource` hits the Self/User API on the backend, which is what data sources are responsible for: requests. 

This repository just takes the responsibilities for the Phone away from the UserRepository as that repository is getting far too big. The repository doesn't need to be tied to phone data sources I don't believe.

File: app/src/main/AndroidManifest.xml (hunk 1 above)
Comment (BradleyWire) [reply app/src/main/AndroidManifest.xml]: Just copied it over from another activity. It isn't needed.

File: app/src/main/kotlin/com/waz/zclient/settings/account/editphonenumber/SettingsAccountPhoneNumberViewModel.kt (hunk 2 above)
Comment (BradleyWire) [reply app/src/main/kotlin/com/waz/zclient/settings/account/editphonenumber/SettingsAccountPhoneNumberViewModel.kt]: The LiveData objects to me represent a different responsibility on the UI the ViewModel can trigger. These all have their own responsibilities based on logic in the ViewModel, I can't make the list smaller without passing some of the logic onto the view.

File: app/src/main/kotlin/com/waz/zclient/user/data/phone/PhoneNumberDataSource.kt (hunk 3 above)
Comment (mejdoo) [reply app/src/main/kotlin/com/waz/zclient/user/data/phone/PhoneNumberDataSource.kt]: Using the same approach can you move responsibilities to some other repositories and get rid of UsersRepository? Why you decided to that only for the PhoneNumber?

Comment (BradleyWire) [reply app/src/main/kotlin/com/waz/zclient/user/data/phone/PhoneNumberDataSource.kt]: Because that's the scope of this PR

Comment (BradleyWire) [reply app/src/main/kotlin/com/waz/zclient/user/data/phone/PhoneNumberDataSource.kt]: Added a TODO (for us in the future) to fix this haha

**Pull request merged**